src/hotspot/share/tsan/tsanOopMap.cpp

```diff
@@ -414,81 +247,86 @@ namespace TsanOopMapImpl {
       guarantee(remaining_moves >= moves_this_cycle,
                 "Excessive number of moves");
       remaining_moves -= moves_this_cycle;
-      DEBUG_PRINT("%s:%d: %d moved, %d remaining\n", __FUNCTION__, __LINE__,
-                  moves_this_cycle, remaining_moves);
+      log_trace(tsan)("%s:%d: %d moved, %d remaining\n", __FUNCTION__, __LINE__,
+                      moves_this_cycle, remaining_moves);
     }
     log_debug(gc)("Tsan: Move %d passes", passes);
   }
+}  // namespace TsanOopMapImpl
 
-  void TsanOopSizeMap::rebuild_oops_map(BoolObjectClosure *is_alive,
-                                        OopClosure *pointer_adjuster) {
-    ResourceMark rm;
-    GrowableArray<PendingMove> moves(MAX2((int)(oop_map->size() / 100),
-                                          100000));
-    bool disjoint_regions;
-    int n_downward_moves;
-    char *min_low, *max_high;
+static OopStorage* _weak_oop_storage;
+static TsanOopMapTable* _oop_map;
 
-    {
-      disjoint_regions = collect_oops(is_alive, pointer_adjuster, &moves,
-                                      &n_downward_moves, &min_low, &max_high);
-    }
-    if (moves.length() == 0) {
-      return;
-    }
+// This is called with TSAN_ONLY, as we want to always create the weak
+// OopStorage so the number matches with the 'weak_count' in oopStorageSet.hpp.
+void TsanOopMap::initialize_map() {
+  // No need to do register_num_dead_callback for concurrent work as we do
```

Comment:
Perhaps we can document this more explicitly, perhaps in tsanOopMap.hpp, that currently TsanOopMap/OopStorage must be processed during a stop-the-world GC pause. This is because we need to update TSAN's metadata for all moved or freed objects during GC, before the mutator reads or writes to these objects. Thus, unlikely other weak OopStorage, TsanOopMap does not support concurrent processing, and does not need to call `OopStorage::register_num_dead_callback()`.

Reply:
SG. I added more details in tsanOopMap.hpp.